mail_tracking/models/mail_message.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016 Antonio Espinosa - <antonio.espinosa@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, api
+
+
+class MailMessage(models.Model):
+    _inherit = "mail.message"
+
+    def _tracking_status_map_get(self):
+        return {
+            'False': 'waiting',
+            'error': 'error',
+            'deferred': 'sent',
+            'sent': 'sent',
+            'delivered': 'delivered',
+            'opened': 'opened',
+            'rejected': 'error',
+            'spam': 'error',
+            'unsub': 'opened',
+            'bounced': 'error',
+            'soft-bounced': 'error',
+        }
+
+    def _partner_tracking_status_get(self, tracking_email):
+        tracking_status_map = self._tracking_status_map_get()
+        status = 'unknown'
+        if tracking_email:
+            tracking_email_status = str(tracking_email.state)
+            status = tracking_status_map.get(tracking_email_status, 'unknown')
+        return status
+
+    def tracking_status(self):
+        res = {}
+        for message in self:
+            partner_trackings = []
+            partners_already = self.env['res.partner']
+            partners = self.env['res.partner']
+            trackings = self.env['mail.tracking.email'].sudo().search([
+                ('mail_message_id', '=', message.id),
+            ])
+            # Search all trackings for this message
+            for tracking in trackings:
+                status = self._partner_tracking_status_get(tracking)
+                recipient = (
+                    tracking.partner_id.name or tracking.recipient)
+                partner_trackings.append((
+                    status, tracking.id, recipient, tracking.partner_id.id))
+                if tracking.partner_id:
+                    partners_already |= tracking.partner_id
+            # Search all recipients for this message
+            if message.partner_ids:
```

Comment:
recordsets operation can be slow compared to normal sets operations. 
You could build a set or a dict ID: name (the only values that you are going to use below)